### DISCUSSION TITLE
Databaseless Email Provider

### DISCUSSION THREAD

**Discussion by giuseppeg** [nextauthjs/next-auth#2526] [Ideas]
The current implementation for the Email Provider stores tokens in the database. Would you consider adding a databaseless/serverless alternative for magic links? It could work as follow:

Create the token and send email:

```js
const email = request.body.email 
const expiration = Date.now() + 6000 * 3 // 3 minutes
const payload = JSON.stringify([email, expiration])
const token = encrypt(SECRET, payload)
const message = `Log in with the following link:

https://acme.com/api/auth/callback/email?token=${code}&email=${email}
`
await sendMail({ to: email, message })
```

API route to handle magic links (`/api/auth/callback/email`):

```js
const { token, email } = request.query
const [tokenEmail, expiration] = JSON.parse(decrypt(SECRET, token))

if (expiration > Date.now() || email !== tokenEmail) { 
  return response.status(401).end()
}

loginOrSignup(tokenEmail)
```

If you think that it is unlikely that you will implement this can you give me some pointers so that I can implement this custom provider myself?

I want this functionality to:

1. Save resources (DB, Serverless functions)
2. Avoid to have cron jobs to clean up the tokens table (remove expired and unused tokens periodically)

**Comment by balazsorban44** [comment]
The same discussion is here: #1913
